avalanche/models/dynamic_optimizers.py

```diff
@@ -16,6 +16,161 @@
+    3: "\033[35m",
+    4: "\033[36m",
+}
+colors[None] = colors["END"]
+
+
+def map_optimized_params(optimizer, new_params):
+    current_parameters_mapping = defaultdict(dict)
+    not_found = []
+    for n, p in new_params.items():
+        g = None
+        # Find param in optimizer
+        found = False
+        for group_idx, group in enumerate(optimizer.param_groups):
+            params = group["params"]
+            for po in params:
+                if id(po) == id(p):
+                    g = group_idx
+                    found = True
+                    break
+        if not found:
+            not_found.append(n)
+        current_parameters_mapping[n] = g
+    return current_parameters_mapping, not_found
+
+
+def build_tree_from_name_groups(name_groups):
+    root = TreeNode("")  # Root node
+    node_mapping = {}
+
+    # Iterate through each string in the list
+    for name, group in name_groups.items():
+        components = name.split(".")
+        current_node = root
+
+        # Traverse the tree and construct nodes for each component
+        for component in components:
+            if component not in current_node.children:
+                current_node.children[component] = TreeNode(
+                    component, parent=current_node
+                )
+            current_node = current_node.children[component]
+
+        # Update the groups for the leaf node
+        if group is not None:
+            current_node.groups |= set([group])
+            current_node.update_groups_upwards()  # Inform parent about the groups
+
+        # Update leaf node mapping dict
+        node_mapping[name] = current_node
+
+    # This will resolve nodes without group
+    root.update_groups_downwards()
+    return root, node_mapping
+
+
+def print_group_information(node, prefix=""):
+    # Print the groups for the current node
+
+    if len(node.groups) == 1:
+        pstring = (
+            colors[list(node.groups)[0]]
+            + f"{prefix}{node.global_name()}: {node.groups}"
+            + colors["END"]
+        )
+        print(pstring)
+    else:
+        print(f"{prefix}{node.global_name()}: {node.groups}")
+
+    # Recursively print group information for children nodes
+    for child_name, child_node in node.children.items():
+        print_group_information(child_node, prefix + "  ")
+
+
+class OptimizedParameterStructure:
+    """
+    Structure holding a tree where each node is a pytorch module
+    the tree is linked to the current model parameters
+    """
+
+    def __init__(self, new_params, optimizer, verbose=False):
+        # Here we rebuild the tree
+        name_groups, not_found = map_optimized_params(optimizer, new_params)
+        self.root, self.node_mapping = build_tree_from_name_groups(name_groups)
+        if verbose:
+            print(f"Not found in optimizer: {not_found}")
+            print_group_information(self.root)
+
+    def __getitem__(self, name):
+        return self.node_mapping[name]
```

Comment:
This is just a class that hold the tree structure and provides a fast mapping from names to leaf nodes

Reply:
I would remove the "Optimized" from the name. `ParameterStructure` is fine. Or `ParameterTree`.